Feature/psalm

mallorydxw opened the pull request.

mallorydxw committed:
Use dxw/result instead of returning array

mallorydxw committed:
Add psalm

mallorydxw committed:
Check for the specific error being thrown

erbridge commented:
Can we get some words about what this is about and for?

erbridge commented on .gitattributes:
```diff
@@ -0,0 +1 @@
+/composer.lock -diff
```
👎

It can be useful to be able to view lock file changes, as they're the record of what's really installed.

mallorydxw replied:
You can still view them with `git diff --text`. This just turns them off by default because they're very long and 99% of the time not what you want to look at.

erbridge replied:
What about `git add -p`? What happens there? I don't think we should assume developers are Git pros. Is there something you can set in your own config to hide things like this?

mallorydxw commented:
> Can we get some words about what this is about and for?

You mean aside from the commit messages? It's adding Psalm (to check for type safety) and fixing one or two bugs Psalm raises.

erbridge commented:
Yes that, thanks.

erbridge commented:
@dxw/govpress-team Could we review this PR and either merge or close it, please?

RobjS committed:
Add a changelog

This project has staged releases, and breaking changes could potentially impact numerous other projects. Therefore we should keep a changelog to help prevent this.

See: https://github.com/dxw/tech-team-rfcs/blob/main/rfc-019-use-changelogs-to-track-changes.md

RobjS committed:
Update README for new Dates::parse return type

It returns a `Dxw\Result\Result` object now, not an array.

RobjS committed:
Update composer dependencies

A psalm dependency was causing issues with Composer 2.0, resulting in failed builds.

RobjS commented:
There's a breaking change here, specifically in https://github.com/dxw/php-missing/pull/17/commits/0cc5c61f0986a56e69ed75cbc082c440df44d39e, where `Dates::parse()` is changed to return a `Dxw\Result\Result` object, rather than an array.

A quick grep of repos I have locally (which I think _should_ cover most that could be using this package) suggests that any using this package are installing `^1.0` as they should be, rather than `dev-master`, so the change shouldn't be an instant problem. But it does make me think we should have a changelog, to help us keep track of this and ensure the next release following this gets a major version bump.

So I've added that, and also updated the README to reflect the new return type.

I've also had to update the composer dependencies, as one of Psalm's dependencies was causing an issue with Composer 2.0. In the course of looking at this, I wanted to `git diff` the `composer.lock` file, which highlighted the potential issue @erbridge raised [above](https://github.com/dxw/php-missing/pull/17#discussion_r458778219). So I've also rebased to remove Mallory's commit to suppress `composer.lock` diffs via `.gitattributes`.

Given I've rebased and updated the dependencies, I probably shouldn't be the one hitting merge on this now. @dxw-serena, please could you review (and merge if all ok) when you have a minute?

serena-piccioni reviewed (approved):
It seems fine.
